README
XSIMPSONS - the Simpsons walking along the tops of your windows

Version 0.1: you get only Bart!

AUTHOR: Stephan Beyer <PH-Linex@gmx.net> 2000

a Simpsons derivate from the original XPenguins...

I would be happy that xsimpsons will become integrated into xpenguins and
with a gtk configuration tool you can set-up if you want penguins or
Simpsons... That would be great!
It may also be included in WinPenguins or made to WinSimpsons.
This software is without any warranty...
!!! Please notice that The Simpsons are created by Matt Groening and
    are the trademarked and copyrighted property of Twentieth Century
    Fox Film Corpoeration...
    I didn't ask for permission, because I didn't found where and why,
    and this is only a free fan software, without commercial aim.
    I take this software away, whenever the Twentieth Century Fox Film
    Corpoeration wants.... !!!


(xpenguins README here)
---

Copyright (C) 1999, 2000  Robin Hogan
Latest version: 1.2 (22 May 2000)


INTRODUCTION

XPenguins is a simple program that animates a number of penguins that
fall from the top of the screen, walk along the top of windows, up
the side of windows, up the side of the screen, and sometimes even
levitate with their genetically-modified go-go-gadget 'copter
ability. Use the `-h' option to see a list of current options -
currently you can specify the display to send the penguins to, the
delay between frames and the number of penguins that are created.


LICENSE

This program is released under the GNU public license - see COPYING


TODO

Write GUI front end(s) - the first will be an applet for GNOME, but
with a little modularisation of the code it should be easy to write
GUIs for other environments (KDE etc) and update them all together. 

Add some more penguin types.

I'd also be interested to see derivatives of this program that use
other toons completely - e.g. Buzz Lightyear, Sonic the Hedgehog, The
Simpsons etc. 


ACKNOWLEDGEMENTS

Inspiration was provided by Rick Jansen <rick@sara.nl> in the form of
the classic `xsnow'.

The penguin images were taken from `Pingus', a free lemmings clone
for Linux (http://pingus.seul.org/), and were designed by Joel Fauche
<joel.fauche@wanadoo.fr>. 

Xpenguins uses vroot.h for use with some `virtual' window managers.
vroot.h is copyright 1991 by Andreas Stolcke, copyright 1990 by
Solbourne Computer Inc. <stolcke@ICSI.Berkeley.EDU>. 


AUTHOR

Robin Hogan <R.J.Hogan@reading.ac.uk>

The xpenguins homepage is  located at: 
	http://www.met.rdg.ac.uk/~swrhgnrj/xpenguins/


SEE ALSO

Michael Vines <isamu@neomueller.org> has rewritten the program for
Microsoft Windows and called it `WinPenguins' - visit: 
	http://neomueller.org/~isamu/winpenguins/ 

    

ChangeLog
* xpenguins 1.2-simpsons1 aka xsimpsons 0.1 (19 June 2000)
- Bart is there
- variant2, too

Original xpenguins 1.2 ChangeLog:
---

* Version 1.2 (22 May 2000)
- Use the X shape extension to allow penguins to negotiate shaped windows
- `-rectwin' option available to bypass calculation of window shape if too slow
- Walker and climber behaviour changed to overcome small obstacles
- Removed whitespace around most penguins images to 30x30 (from 32x32)
- Walker image changed so walker actually on window, not 1 pixel above!
- Fixed bug in ToonAdvance that caused toons to get stuck behind KDE panel
- Renamed functions in toon.c so that all begin with `Toon'
- Tumblers now accelerate under gravity until they reach `terminal velocity'
- Can use `-n' as well as `-penguins' to specify the number of penguins
- Climbers maintain tendency to climb until they next become faller/tumbler
- Every timestep perform check that penguins are not in the window region
- Walkers and climbers now move with the windows they're on (up to a limit)
- Penguins fall in properly from the very top of the screen

* Version 1.1 (26 February 2000)
- Use Xpm attribute `closeness' so that 8-bit displays are supported
- Smoothed the jagged edges of floater image (from Pingus)
- Added the `-ignorepopups', `-version' and `-quiet' options
- Second interupt causes bursting penguins to be immediately erased
- Fixed ToonLocateWindows bug that returned before all windows mapped

* Version 1.0 (28 November 1999)
- Decided that version numbers less than 1 are silly



toon.h
/* toon.h - functions for drawing cartoons on the root window - used by xpenguins
 * Copyright (C) 1999, 2000  Robin Hogan
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include <X11/Intrinsic.h>
#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/Xos.h>
#include <X11/xpm.h>
#include <X11/extensions/shape.h>

#define TOON_UNASSOCIATED -2
#define TOON_HERE -1
#define TOON_LEFT 0
#define TOON_RIGHT 1
#define TOON_UP 2
#define TOON_DOWN 3

#define TOON_FORCE 1
#define TOON_MOVE 0
#define TOON_STILL -1

#define TOON_OK 1
#define TOON_PARTIALMOVE 0
#define TOON_BLOCKED -1
#define TOON_SQUASHED -2

#define TOON_DEFAULTS 0L
#define TOON_NOEDGEBLOCK (1L<<0)
#define TOON_EDGEBLOCK (1L<<1)
#define TOON_SIDEBOTTOMBLOCK (1L<<2)
#define TOON_NOSOLIDPOPUPS (1L<<4)
#define TOON_SOLIDPOPUPS (1L<<5)
#define TOON_NOSHAPEDWINDOWS (1L<<6)
#define TOON_SHAPEDWINDOWS (1L<<7)

#define TOON_NOCYCLE (1L<<8)

#define TOON_NOCATCHSIGNALS (1L<<16)
#define TOON_CATCHSIGNALS (1L<<17)
#define TOON_EXITGRACEFULLY (1L<<18)

#define TOON_MESSAGE_LENGTH 64
#define TOON_DEFAULTMAXRELOCATE 8

/*** STRUCTURES ***/

typedef struct {
   char
      **image;
   int
      nframes,ndirections, /* number of frames and directions in image */
      width,height; /* width and height of an individual frame/direction */
   long int
      conf; /* bitmask of toon properties such as cycling etc. */
   Pixmap
      pixmap, mask; /* pointers to X structures */
} ToonData;


typedef struct {
   int
      x,y,u,v, /* new position and velocity */
      active,type,frame,direction,
      x_map,y_map,width_map,height_map,
            /* properties of the image mapped on the screen */
      associate, /* toon is associated with a window */
      xoffset, yoffset; /* location relative to window origin */
   unsigned int wid; /* window associated with */   
} Toon;

/*** FUNCTION PROTOTYPES ***/

/* SIGNAL AND ERROR HANDLING FUNCTIONS */
int ToonSignal();
char *ToonErrorMessage();

/* STARTUP FUNCTIONS */
Display *ToonOpenDisplay(char *display_name);
int ToonConfigure(unsigned long int code);
int ToonInstallData(ToonData *toon_data, int n);

/* DRAWING FUNCTIONS */
int ToonDraw(Toon *toon,int n);
int ToonErase(Toon *toon,int n);
void ToonFlush();

/* QUERY FUNCTIONS */
int ToonDisplayWidth();
int ToonDisplayHeight();
int ToonBlocked(Toon *toon, int direction);
int ToonOffsetBlocked(Toon *toon, int xoffset, int yoffset);
int ToonWindowsMoved();

/* ASSIGNMENT FUNCTIONS */
void ToonMove(Toon *toon, int xoffset, int yoffset);
void ToonSetPosition(Toon *toon, int x, int y);
void ToonSetType(Toon *toon, int type, int direction, int gravity);
void ToonSetVelocity(Toon *toon, int u, int v);

/* CORE FUNCTIONS */
int ToonAdvance(Toon *toon, int mode);
int ToonLocateWindows();
int ToonSleep(unsigned long usecs);

/* FINISHING UP */
int ToonCloseDisplay();

/* HANDLING TOON ASSOCIATIONS WITH MOVING WINDOWS */
void ToonSetAssociation(Toon *toon, int direction);
void ToonSetMaximumRelocate(int up, int down, int left, int right);
int ToonRelocateAssociated(Toon *toon, int n);
int ToonCalculateAssociations(Toon *toon, int n);



vroot.h
/*****************************************************************************/
/**                   Copyright 1991 by Andreas Stolcke                     **/
/**               Copyright 1990 by Solbourne Computer Inc.                 **/
/**                          Longmont, Colorado                             **/
/**                                                                         **/
/**                           All Rights Reserved                           **/
/**                                                                         **/
/**    Permission to use, copy, modify, and distribute this software and    **/
/**    its documentation  for  any  purpose  and  without  fee is hereby    **/
/**    granted, provided that the above copyright notice appear  in  all    **/
/**    copies and that both  that  copyright  notice  and  this  permis-    **/
/**    sion  notice appear in supporting  documentation,  and  that  the    **/
/**    name of Solbourne not be used in advertising                         **/
/**    in publicity pertaining to distribution of the  software  without    **/
/**    specific, written prior permission.                                  **/
/**                                                                         **/
/**    ANDREAS STOLCKE AND SOLBOURNE COMPUTER INC. DISCLAIMS ALL WARRANTIES **/
/**    WITH REGARD TO THIS SOFTWARE, INCLUDING ALL IMPLIED WARRANTIES OF    **/
/**    MERCHANTABILITY  AND  FITNESS,  IN  NO  EVENT SHALL ANDREAS STOLCKE  **/
/**    OR SOLBOURNE BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL    **/
/**    DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA   **/
/**    OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER    **/
/**    TORTIOUS ACTION, ARISING OUT OF OR IN  CONNECTION  WITH  THE  USE    **/
/**    OR PERFORMANCE OF THIS SOFTWARE.                                     **/
/*****************************************************************************/
/*
 * vroot.h -- Virtual Root Window handling header file
 *
 * This header file redefines the X11 macros RootWindow and DefaultRootWindow,
 * making them look for a virtual root window as provided by certain `virtual'
 * window managers like swm and tvtwm. If none is found, the ordinary root
 * window is returned, thus retaining backward compatibility with standard
 * window managers.
 * The function implementing the virtual root lookup remembers the result of
 * its last invocation to avoid overhead in the case of repeated calls
 * on the same display and screen arguments. 
 * The lookup code itself is taken from Tom LaStrange's ssetroot program.
 *
 * Most simple root window changing X programs can be converted to using
 * virtual roots by just including
 *
 * #include <X11/vroot.h>
 *
 * after all the X11 header files.  It has been tested on such popular
 * X clients as xphoon, xfroot, xloadimage, and xaqua.
 * It also works with the core clients xprop, xwininfo, xwd, and editres
 * (and is necessary to get those clients working under tvtwm).
 * It does NOT work with xsetroot; get the xsetroot replacement included in
 * the tvtwm distribution instead.
 *
 * Andreas Stolcke <stolcke@ICSI.Berkeley.EDU>, 9/7/90
 * - replaced all NULL's with properly cast 0's, 5/6/91
 * - free children list (suggested by Mark Martin <mmm@cetia.fr>), 5/16/91
 * - include X11/Xlib.h and support RootWindowOfScreen, too 9/17/91
 */

#ifndef _VROOT_H_
#define _VROOT_H_

//#if !defined(lint) && !defined(SABER)
//    static char vroot_rcsid[] = "$Id: vroot.h,v 1.4 1991/09/30 19:23:16 stolcke Exp stolcke $";
//#endif

#include <X11/X.h>
#include <X11/Xatom.h>
#include <X11/Xlib.h>

static Window
VirtualRootWindowOfScreen(screen)
	Screen *screen;
{
	static Screen *save_screen = (Screen *)0;
	static Window root = (Window)0;

	if (screen != save_screen) {
		Display *dpy = DisplayOfScreen(screen);
		Atom __SWM_VROOT = None;
		int i;
		Window rootReturn, parentReturn, *children;
		unsigned int numChildren;

		root = RootWindowOfScreen(screen);

		/* go look for a virtual root */
		__SWM_VROOT = XInternAtom(dpy, "__SWM_VROOT", False);
		if (XQueryTree(dpy, root, &rootReturn, &parentReturn,
				 &children, &numChildren)) {
			for (i = 0; i < numChildren; i++) {
				Atom actual_type;
				int actual_format;
				unsigned long nitems, bytesafter;
				Window *newRoot = (Window *)0;

				if (XGetWindowProperty(dpy, children[i],
					__SWM_VROOT, 0, 1, False, XA_WINDOW,
					&actual_type, &actual_format,
					&nitems, &bytesafter,
					(unsigned char **) &newRoot) == Success
				    && newRoot) {
				    root = *newRoot;
				    break;
				}
			}
			if (children)
				XFree((char *)children);
		}

		save_screen = screen;
	}

	return root;
}

#undef RootWindowOfScreen
#define RootWindowOfScreen(s) VirtualRootWindowOfScreen(s)

#undef RootWindow
#define RootWindow(dpy,screen) VirtualRootWindowOfScreen(ScreenOfDisplay(dpy,screen))

#undef DefaultRootWindow
#define DefaultRootWindow(dpy) VirtualRootWindowOfScreen(DefaultScreenOfDisplay(dpy))

#endif /* _VROOT_H_ */


variant2/def.h
/* def.h - defines penguin image data for use by xpenguins
 * Copyright (C) 1999, 2000  Robin Hogan
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#define PENGUIN_FORWARD 0
#define PENGUIN_LEFTRIGHT 1
#define PENGUIN_LEFT 0
#define PENGUIN_RIGHT 1

#define PENGUIN_WALKER 0
#define PENGUIN_FALLER 1
#define PENGUIN_TUMBLER 2
#define PENGUIN_FLOATER 3
#define PENGUIN_CLIMBER 4
#define PENGUIN_BOMBER 5
#define PENGUIN_EXPLOSION 6
#define PENGUIN_TYPES 7

#define PENGUIN_DEFAULTWIDTH 30
#define PENGUIN_DEFAULTHEIGHT 30

#include "walker.xpm"
#include "faller.xpm"
#include "tumbler.xpm"
#include "floater.xpm"
#include "climber.xpm"
#include "bomber.xpm"
#include "explosion.xpm"

ToonData penguin_data[] = {
   { walker_xpm , 1, 2, 42, 61, TOON_DEFAULTS },
   { faller_xpm , 4, 1, 46, 62, TOON_DEFAULTS },
   { tumbler_xpm, 1, 1, 46, 62, TOON_DEFAULTS },
   { floater_xpm, 4, 1, 46, 62, TOON_DEFAULTS },
   { climber_xpm, 1, 2, 61, 43, TOON_DEFAULTS },
   { bomber_xpm , 1, 1, 64, 64, TOON_NOCYCLE },
   { explosion_xpm, 1, 1, 64, 64, TOON_NOCYCLE }
};


xsimpsons.c
/* XPENGUINS - cool little penguins that walk along the tops of your windows
 * Copyright (C) 1999, 2000  Robin Hogan
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 *
 * XPENGUINS is a kind of a cross between lemmings and xsnow - penguins
 * fall from the top of the screen and walk across the tops of windows.
 * The images are take from `Pingus', a lemmings-type game for Linux.
 * All the nasty X calls are in `toon.c', so xpenguins.c should be easily
 * modifiable to animate any small furry (or feathery) animal of your choice.
 * 
 * Robin Hogan <R.J.Hogan@reading.ac.uk>
 * Project started 22 November 1999
 * Last update 26 February 2000
 */

#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>

/* C structures defined here */
#include "toon.h"

/*
 * The penguin shapes are defined here - includes links to all the xpm images.
 * If you want to have something else falling from the top of your screen (the
 * possibilities are endless!) then point the following line somewhere else. 
 */
#include "penguins/def.h"

#define MAX_PENGUINS 256
#define DEFAULT_DELAY 50
#define JUMP_DISTANCE 8
#define RandInt(maxint) ((int) ((maxint)*((float) rand()/(RAND_MAX+1.0))))

#define XPENGUINS_VERSION "1.2"
#define XPENGUINS_AUTHOR "Robin Hogan"
#define XPENGUINS_DATE "22 May 2000"

/* A few macros... */

#define InitPenguin(penguin) \
   ToonSetType(penguin, PENGUIN_FALLER, PENGUIN_FORWARD, \
         TOON_UNASSOCIATED); \
   ToonSetPosition(penguin, RandInt(ToonDisplayWidth() \
         - PENGUIN_DEFAULTWIDTH), 1-PENGUIN_DEFAULTHEIGHT); \
   ToonSetAssociation(penguin, TOON_UNASSOCIATED); \
   ToonSetVelocity(penguin, RandInt(2)*2-1, 3)

#define MakeClimber(penguin) \
   ToonSetType(penguin, PENGUIN_CLIMBER, (penguin)->direction, \
         TOON_DOWN); \
   ToonSetAssociation(penguin, (penguin)->direction); \
   ToonSetVelocity((penguin),0,-4)

#define MakeWalker(penguin) \
   ToonSetType(penguin, PENGUIN_WALKER, \
         (penguin)->direction,TOON_DOWN); \
   ToonSetAssociation(penguin, TOON_DOWN); \
   ToonSetVelocity(penguin, 4*((2*(penguin)->direction)-1), 0)

#define MakeFaller(penguin) \
   ToonSetVelocity(penguin, ((penguin)->direction)*2-1, 3); \
   ToonSetType(penguin, PENGUIN_FALLER, \
         PENGUIN_FORWARD,TOON_DOWN); \
   ToonSetAssociation(penguin, TOON_UNASSOCIATED)

void ShowUsage(char **argv) {
   fprintf(stdout,"Usage: %s [options]\n",argv[0]);
   fprintf(stdout,"Options:\n");
   fprintf(stdout,"  -display <display>        Send the penguins to <display>'\n");
   fprintf(stdout,"  -delay <millisecs>        Set delay between frames (default %d)\n",
         DEFAULT_DELAY);
   fprintf(stdout,"  -n, -penguins <n>         Create <n> penguins (max %d)\n",
         MAX_PENGUINS);
   fprintf(stdout,"  -ignorepopups             Penguins ignore `popup' windows\n");
   fprintf(stdout,"  -rectwin                  Regard shaped windows as rectangular\n");
   fprintf(stdout,"  -q, -quiet, --quiet       Suppress message on exit\n");
   fprintf(stdout,"  -v, -version, --version   Show version information\n");
   fprintf(stdout,"  -h, -?, -help, --help     Show this message\n");
};


/* Global variables */
int finished=0;
int new_positions=0;
int verbose=1;

/*** MAIN PROGRAM ***/
int main (int argc, char **argv) {
   unsigned long sleep_usec=DEFAULT_DELAY*1000;
   unsigned long configure_mask = TOON_SIDEBOTTOMBLOCK 
         | TOON_CATCHSIGNALS;
   char *display_name=NULL;
   int windows_moved,status,npenguins=8,i,n,direction;
   Toon penguin[MAX_PENGUINS];
   char prefd[MAX_PENGUINS]; /* preferred direction; -1 means none */
   char prefclimb[MAX_PENGUINS]; /* climbs when possible */
   char hold_on[MAX_PENGUINS];
   /* Handle command-line arguments */
   for (n=1;n<argc;n++) {
      if (strcmp(argv[n],"-n") == 0 || strcmp(argv[n],"-penguins") == 0) {
         if (argc > ++n) {
            npenguins=atoi(argv[n]);
            if (npenguins > MAX_PENGUINS) {
               fprintf(stderr,"Warning: only %d penguins created\n",
                     npenguins=MAX_PENGUINS);
            }
            else if (npenguins <= 0) {
               fprintf(stderr,"Warning: no penguins created\n");
                     npenguins=0;
            }
         }
         else {
            fprintf(stderr,"Error: number of penguins not specified\n");
            ShowUsage(argv);
            exit(1);
         }
      }
      else if (strcmp(argv[n],"-delay") == 0) {
         if (argc > ++n) {
            sleep_usec=1000*atoi(argv[n]);
         }
         else {
            fprintf(stderr,"Error: delay in milliseconds not specified\n");
            ShowUsage(argv);
            exit(1);
         }
      }
      else if (strcmp(argv[n],"-display") == 0) {
         if (argc > ++n) {
            display_name=argv[n];
         }
         else {
            fprintf(stderr,"Error: display not specified\n");
            ShowUsage(argv);
            exit(1);
         }
      }
      else if (strcmp(argv[n],"-ignorepopups") == 0 ) {
         configure_mask |= TOON_NOSOLIDPOPUPS;
      }
      else if (strcmp(argv[n],"-rectwin") == 0 ) {
         configure_mask |= TOON_NOSHAPEDWINDOWS;
      }
      else if (strcmp(argv[n],"-h") == 0 || strcmp(argv[n],"-help") == 0 || 
               strcmp(argv[n],"-?") == 0 || strcmp(argv[n],"--help") == 0) {
         fprintf(stdout,"XPenguins %s (%s) by %s\n",
               XPENGUINS_VERSION, XPENGUINS_DATE, XPENGUINS_AUTHOR);
         ShowUsage(argv);
         exit(0);
      }
      else if (strcmp(argv[n],"-v") == 0 || strcmp(argv[n],"-version") == 0 || 
               strcmp(argv[n],"--version") == 0) {
         fprintf(stdout,"XPenguins %s\n", XPENGUINS_VERSION);
         exit(0);
      }
      else if (strcmp(argv[n],"-q") == 0 || strcmp(argv[n],"-quiet") == 0 || 
               strcmp(argv[n],"--quiet") == 0) {
         verbose=0;
      }
      else {
         fprintf(stderr,"Warning: `%s' not understood - use `-h' for a list of options\n",argv[n]);
      }
   }

   /* reset random-number generator */
   srand(time((long *) NULL));

   /* contact X server and set up some basic X stuff */
   if (ToonOpenDisplay(display_name) == NULL) {
      fprintf(stderr,"Error: %s\n", ToonErrorMessage());
      exit(1);
   };
   /* Set up various preferences: Edge of screen is solid, and if a signal is caught
    * then exit the main event loop */
   ToonConfigure(configure_mask);
   /* Set the distance the window can move (up, down, left, right) and penguin
    * can still cling on */
   ToonSetMaximumRelocate(16,16,16,16);
   /* Send the pixmaps to the X server  - penguin_data should have been 
    * defined in penguins/def.h */
   ToonInstallData(penguin_data,PENGUIN_TYPES);

   /* initialise penguins */
   for (i=0;i<npenguins;i++) {
      prefd[i] = -1;
      prefclimb[i] = 0;
      hold_on[i] = 0;
      InitPenguin(penguin+i);
   }

   /* Find out where the windows are - should be done just before beginning the 
    * event loop */
   ToonLocateWindows();
   /* Event loop */
   while (!finished) {
      /* check if windows have moved, and flush the display */
      if ( (windows_moved = ToonWindowsMoved()) ) {
         /* if so, check for squashed toons */
         ToonCalculateAssociations(penguin,npenguins);
         windows_moved = ToonLocateWindows();
         ToonRelocateAssociated(penguin,npenguins);
      }
      for (i=0;i<npenguins;i++) {
         if (!penguin[i].active) {
            InitPenguin(penguin+i);
            continue;
         }
         else {
            if (ToonBlocked(penguin+i,TOON_HERE)) {
               ToonSetType(penguin+i,PENGUIN_EXPLOSION,
                     PENGUIN_FORWARD,TOON_HERE);
               ToonSetAssociation(penguin+i, TOON_UNASSOCIATED);
            }

            status=ToonAdvance(penguin+i,TOON_MOVE);
            switch (penguin[i].type) {
               case PENGUIN_FALLER:
                  if (status != TOON_OK) {
                     if (ToonBlocked(penguin+i,TOON_DOWN)) {
                        if (prefd[i]>-1)
                           penguin[i].direction=prefd[i];
                        else
                           penguin[i].direction=RandInt(2);
                        MakeWalker(penguin+i);
                        prefd[i]=-1;
                     }
                     else {
                        if (RandInt(2)) {
                           ToonSetVelocity(penguin+i,-penguin[i].u,3);
                        }
                        else {
                           penguin[i].direction = penguin[i].u>0;
                           MakeClimber(penguin+i);
                        }
                     }
                  }
                  break;

               case PENGUIN_TUMBLER:
                  if (status != TOON_OK) {
                     if (prefd[i]>-1)
                        penguin[i].direction=prefd[i];
                     else
                        penguin[i].direction=RandInt(2);
                     MakeWalker(penguin+i);
                     prefd[i]=-1;
                  }
                  else if (penguin[i].v < 8) {
                     penguin[i].v +=1;
                  }
                  break;

               case PENGUIN_WALKER:
                  if (status != TOON_OK) {
                     if (status == TOON_BLOCKED) {
                        /* Try to step up... */
                        int u = penguin[i].u;
                        if (!ToonOffsetBlocked(penguin+i, u, -JUMP_DISTANCE)) {
                           ToonMove(penguin+i, u, -JUMP_DISTANCE);
                           ToonSetVelocity(penguin+i, 0, JUMP_DISTANCE-1);
                           ToonAdvance(penguin+i, TOON_MOVE);
                           ToonSetVelocity(penguin+i, u, 0);
                        }
                        else {
                           /* Blocked! We can turn round, fly or climb... */
                           switch (RandInt(8)*(1-prefclimb[i])) {
                              case 0:
                                 MakeClimber(penguin+i);
                                 break;
                              case 1:
                                 ToonSetType(penguin+i,PENGUIN_FLOATER,
                                       PENGUIN_FORWARD,TOON_DOWN);
                                 ToonSetAssociation(penguin+i, TOON_UNASSOCIATED);
                                 ToonSetVelocity(penguin+i,RandInt(5)
                                       * (-penguin[i].u/4),-3);
                                 break;
                              default:
                                 penguin[i].direction = (!penguin[i].direction);
                                 MakeWalker(penguin+i);
                           }
                        }
                     }
                  }
                  else if (!ToonBlocked(penguin+i,TOON_DOWN)) {
                     /* Try to step down... */
                     ToonSetVelocity(penguin+i, 0, JUMP_DISTANCE);
                     status=ToonAdvance(penguin+i,TOON_MOVE);
                     if (status == TOON_OK) {
                        prefd[i]=penguin[i].direction;
                        ToonSetType(penguin+i, PENGUIN_TUMBLER,
                              PENGUIN_FORWARD,TOON_DOWN);
                        ToonSetAssociation(penguin+i, TOON_UNASSOCIATED);
                        ToonSetVelocity(penguin+i, 0, 1);
                        prefclimb[i]=0;
                     }
                     else {
                        ToonSetVelocity(penguin+i, 4*((2*penguin[i].direction)-1), 0);
                     }
                  }
                  break;

               case PENGUIN_CLIMBER:
                  direction = penguin[i].direction;
                  if (penguin[i].y < 0) {
                     penguin[i].direction = (!direction);
                     MakeFaller(penguin+i);
                     prefclimb[i]=0;
                  }
                  else if (status == TOON_BLOCKED) {
                     /* Try to step out... */
                     int v = penguin[i].v;
                     int xoffset = (1-direction*2) * JUMP_DISTANCE;
                     if (!ToonOffsetBlocked(penguin+i, xoffset, v)) {
                        ToonMove(penguin+i, xoffset, v);
                        ToonSetVelocity(penguin+i, -xoffset-(1-direction*2), 0);
                        ToonAdvance(penguin+i, TOON_MOVE);
                        ToonSetVelocity(penguin+i, 0, v);
                     }
                     else {
                        penguin[i].direction = (!direction);
                        MakeFaller(penguin+i);
                        prefclimb[i]=0;
                     }
                  }
                  else if (!ToonBlocked(penguin+i,direction)) {
                     if (ToonOffsetBlocked(penguin+i, ((2*direction)-1)
                           * JUMP_DISTANCE, 0)) {
                        ToonSetVelocity(penguin+i, ((2*direction)-1)
                           * (JUMP_DISTANCE-1), 0);
                        ToonAdvance(penguin+i, TOON_MOVE);
                        ToonSetVelocity(penguin+i, 0, -4);
                     }
                     else {
                        MakeWalker(penguin+i);
                        ToonSetPosition(penguin+i, penguin[i].x+(2*direction)-1,
                              penguin[i].y);
                        prefd[i]=direction;
                        prefclimb[i]=1;
                     }
                  }
                  break;

               case PENGUIN_FLOATER:
                  if (penguin[i].y < 0) {
                     penguin[i].direction = (penguin[i].u>0);
                     MakeFaller(penguin+i);
                  }
                  else if (status != TOON_OK) {
                     if (ToonBlocked(penguin+i,TOON_UP)) {
                       penguin[i].direction = (penguin[i].u>0);
                       MakeFaller(penguin+i);
                     }
                     else {
                        ToonSetVelocity(penguin+i,-penguin[i].u, -3);
                     }
                  }
                  break;

               case PENGUIN_EXPLOSION:
                  if (!hold_on[i]) {
                     hold_on[i] = 1;
                  }
                  else {
                     penguin[i].active=0;
                     hold_on[i]=0;
                  }
             }
         }
      }
      /* First erase them all, then draw them all - should reduce flickering */
      ToonErase(penguin,npenguins);
      ToonDraw(penguin,npenguins);
      ToonFlush();
      /* pause */
      ToonSleep(sleep_usec);
      /* Has an interupt signal been received? If so, quit gracefully */
      finished=ToonSignal();
   }

   /* Exit sequence... */
   /* Any more signals (TERM, HUP or INT) and the penguins are 
    * erased immediately */
   ToonConfigure(TOON_EXITGRACEFULLY);
   if (1) {
      /* Nice exit sequence... */
      if (verbose) fprintf(stderr,"Interupt received: exploding penguins");
      for (i=0;i<npenguins;i++) {
         if (penguin[i].active) {
            ToonSetType(penguin+i,PENGUIN_BOMBER,
                  PENGUIN_FORWARD,TOON_DOWN);
         }
      }
      for (n=0;n<penguin_data[PENGUIN_BOMBER].nframes;n++) {
         ToonErase(penguin,npenguins);
         ToonDraw(penguin,npenguins);
         ToonFlush();
         for (i=0;i<npenguins;i++) {
            ToonAdvance(penguin+i,TOON_FORCE);
         }
         ToonSleep(sleep_usec);
         if (verbose) fprintf(stderr,".");
      }
      if (verbose) fprintf(stderr,"done\n");
   }
   ToonErase(penguin,npenguins);
   ToonCloseDisplay();
   exit(0);
}

